File: java/programmers/brute-force-search/숫자야구.java
/*
* 🙆‍♂️ Created by wwlee94 on 2020.02.26
https://programmers.co.kr/learn/courses/30/lessons/42841

- Java JDK 8 버전에서 Collection to Array, Array to Collection 변환하기 ! -
* Collection to Array
    ArrayList<Integer> choosed = new ArrayList<>();
    1. Wrapper Class Array로 만들 때
        Integer[] myArray = choosed.stream().toArray(Integer[]::new);
    2. Primitive Array로 만들 때
        int[] myArray = choosed.stream().mapToInt(x->x).toArray();
* Array to Collection
    1. Wrapper Class Array일 때
        Integer[] myArray = new Integer[10];
        ArrayList<Integer> list = Arrays.stream(myArray).collect(Collectors.toList());
    2. Primitive Array일 때
        int[] myArray = new int[10];
        ArrayList<Integer> list = Arrays.stream(myArray).boxed().collect(Collectors.toList());

- ArrayList<int[]> 출력하기 -
    for(int[] myInt: all_case) System.out.println(Arrays.toString(myInt));

- 다른 풀이 방법 -
1. question을 구할 때 int값을 String으로 변환한 후 str[0]- 백의 자리, str[1] - 십의자리 ... 이런식으로 구하는 방법도 있음
2. ball을 구할때 cases와 question의 교집합에서 strike 개수를 빼서 구하는 방법도 있음
*/
import java.util.*;
class Solution {
    int[] used;
    ArrayList<Integer> choosed = new ArrayList<>();
    ArrayList<int[]> all_case = new ArrayList<>();
    
    public int solution(int[][] baseball) {
        int answer = 0;
        int[] number = {1,2,3,4,5,6,7,8,9};
        permutations(number, 3);
        
        // for(int[] myInt: all_case) System.out.println(Arrays.toString(myInt));
        for (int[] cases: all_case){
            int count = 0; //질문을 만족한 개수 (strike, ball)
            for (int[] item: baseball){
                int strike = 0;
                int ball = 0;
                int[] question = {item[0] / 100, (item[0] % 100) / 10, item[0] % 10};

                for(int i=0;i<3;i++){
                    for(int j=0;j<3;j++){
                        if(cases[i] == question[j]){
                            if(i==j){ strike += 1; }
                            else{ ball += 1; }
                        }
                    }
                }
                if (item[1] != strike || item[2] != ball) break;
                else count += 1;
            }
            if (count == baseball.length) answer += 1;
        }
        return answer;
    }
    
    // 자바는 순열 직접 구현 해야함
    public void permutations(int[] array, int r){
        used = new int[array.length]; //자동 0으로 초기화
        generate(array, r);
    }
    
    public void generate(int[] array, int r){
        if (choosed.size() == r) {
            // Integer[] myArray = choosed.stream().toArray(Integer[]::new);
            int[] myArray = choosed.stream().mapToInt(x->x).toArray();
            all_case.add(myArray);
            return;
        }
        
        for(int i=0;i<array.length;i++){
            if (used[i] == 0){
                used[i] = 1;
                choosed.add(array[i]);
                generate(array, r);
                used[i] = 0;
                choosed.remove(choosed.size()-1); // Remove last element
            }
        }
    }
}

File: python/programmers/brute-force-search/숫자야구.py
'''
* 🙆‍♂️ Created by wwlee94 on 2020.02.26
https://programmers.co.kr/learn/courses/30/lessons/42841?language=python3

- 문제 풀이 접근법 -
1. 문제에서 주어지는 질문과 일치하는 숫자를 찾는 것이기 때문에 1~9 사이 숫자로 만들 수 있는 3자리 순열을 생성함
2. 주어진 질문과 비교해서 strike, ball을 구함
3. 어떠한 3자리 수가 주어진 질문과 모두 일치 한다면 그 숫자는 정답 가능성이 있는 숫자이므로 answer += 1 해줌 !

- 기존 풀이 -
1. question을 구하는 다른 방법 (But, 자리 수가 늘어나는 경우엔 작성하기 귀찮아짐)
question = [item[0] // 100, item[0] % 100 // 10, item[0] % 10]

2. strike, ball을 구하는 다른 방법 (But, 첫번째 방법보다 조금 비효율적?)
# 2중 for문으로 strike, ball 구한 것 
    for i in range(3):
        for j in range(3):
            if i==j and case[i] == question[j]: strike += 1
            elif case[i] == question[j]: ball += 1
    if item[1] != strike or item[2] != ball: break
    else: count += 1
'''
import itertools

def solution(baseball):
    answer = 0
    leng = len(baseball)
    all_case = list(itertools.permutations([1,2,3,4,5,6,7,8,9], 3))
    
    for case in all_case:
        count = 0 # 질문을 만족한 개수
        for item in baseball:
            strike = 0
            question = [int(i) for i in str(item[0])]

            #첫번째 방법 - strike 개수 우선 구한 후 ball 개수 구함
            for i in range(3):
                if case[i] == question[i]: strike += 1
            if item[1] != strike: break
                
            ball = len(set(case) & set(question)) - strike # 교집합의 크기를 구한 후 strike 개수 빼준 것
            if item[2] != ball: break
                
            count += 1
                
        if count == leng: answer += 1 # 질문 개수 만큼 돌아야 조건을 모두 만족한 경우이기 때문
    return answer